formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt

```diff
@@ -200,20 +247,98 @@ internal open class ProtobufDecoder(
 
     override fun SerialDescriptor.getTag(index: Int) = extractParameters(index)
 
+    private fun findUnreadElementIndex(): Int {
+        val elementsCount = descriptor.elementsCount
+        while (lowerReadMark != -1L) {
+            val index = lowerReadMark.inv().countTrailingZeroBits()
+            lowerReadMark = lowerReadMark or (1L shl index)
+
+            if (!descriptor.isElementOptional(index)) {
+                val elementDescriptor = descriptor.getElementDescriptor(index)
```

Comment:
Is this possible to evaluate these conditions beforehand? So we won't iterate the whole bitmask every time

Probably we can mark every non-optional and nullable(or collection) field with one bits and others with zeroes in `prepareReadMarks` (read values would be zero too). Then it's possible to use countLeading/TrailingZeroBits right away without need to retrieve element descriptor

Although I'm not sure if such optimization would really make the difference in benchmarks. If you have strong objections, feel free to leave this as is

Reply:
Evaluate these conditions beforehand will force to always read descriptor's elements properties - this will slow down the execution for cases with all required fields.

Reply:
`getElementDescriptor` is a simple O(1) array access. But ok, note taken